### ISSUE TITLE
Possibly bug with Params editorconfig handling

### ISSUE THREAD

**Issue by ZacSweers** [pinterest/ktlint#605]
I've been working on adding editorconfig support to the Spotless gradle plugin, but I believe I'm hitting an issue due to this line:

https://github.com/pinterest/ktlint/blob/master/ktlint-core/src/main/kotlin/com/pinterest/ktlint/core/KtLint.kt#L178

If the input file name doesn't match (such as if it's null), it will ignore the editorconfig that was passed in and return an empty map. My sense is that this was an oversight and it should just return the parsed mapping just without the `file_name` added to it. Before I make a PR I wanted to confirm if this was the case, or understand better why if it's intentional.

This presents an issue for spotless because it doesn't currently pass in the file name to the format call, just the contents. It might be possible to change this if need be (and maybe should be the case for other rules that depend on `file_name`), but that's the current reason this can't be simply worked around.

**Comment by Tapchicoma** [comment]
Before moving `.editorconfig` processing to core package this piece of code looked following:
```kotlin
if (editorConfigPath != null) {
    val userData = (
        EditorConfig.of(File(editorConfigPath).canonicalPath)
            ?.onlyIf({ debug }) { printEditorConfigChain(it) }
            ?: emptyMap<String, String>()
        ) + cliUserData
    return fun (fileName: String) = userData + ("file_path" to fileName)
}
```
and `fileName` in ktlint cli is always no-null (stdin has special value). Though in ktlint-core `fileName` param is nullable. Changing it to be non-nullable will be a breaking change. Not sure if mapping without `fileName` could be used as a solution.

**Comment by paul-dingemans** [comment]
Most likely this issue is obsolete due to changes in ktlint `0.46.x`.

**Comment by paul-dingemans** [comment]
@ZacSweers How relevant is your question at this moment? I have been reading the [spotless issue](https://github.com/diffplug/spotless/issues/142). If I understand correctly, the goal is that when Spotless invokes `ktlint` that `.editorconfig` is supported by ktlint.

As you remarked in [comment](https://github.com/diffplug/spotless/issues/142#issuecomment-539171202), ktlint needs to know the path of the file so that it can check which `.editorconfig` files are located on the path to that file. The files are combined hierarchally where files deeper in the hierarchy are more important.

In case `ExperimentalParams.editorConfigPath` is specified, all '.editorconfig' files on the path to the file are being ignored. Only the setting from the `ExperimentalParams.editorConfigPath` are read and used.

I am open for suggestions to improve the API of ktlint but at this moment I can not figure out what Spotless's needs are given the current state of both Spotless and KtLint.

One possibily could be that the singular `ExperimentalParams.editorConfigPath` is changed to a list of paths to `.editorConfig` where files are ranked from most important to least important. This would allow spotless to collect and track the status of the files.

Another possibility is that an API is added which, given a directory path creates an instance of the EditorConfigOverride based on the '.editorConfig' files which then can be supplierd to calls of `'lint` and `format` for all files in the directory.

**Comment by paul-dingemans** [comment]
> Sorry I no longer use ktlint, so I don't think I have any helpful input

Ok, tnx for replying.

I have updated the spotless issue to inform other participants about status in ktlint. Closing this issue for now.
